src/utils/dom-ancestor-visitor.js

```diff
@@ -0,0 +1,143 @@
+ * their maximum nodes to visit, no more nodes will be visited.
+ *
+ * Example usage:
+ * new DomAncestorVisitor()
+ *   .addVisitor((el, style) => { ... })
+ *   .addVisitor((el, style) => { ... })
+ *   ...
+ *   .visitAncestorsStartingFrom(someElement);
+ */
+export class DomAncestorVisitor {
+  /** @param {!Window=} win */
+  constructor(win = window) {
+    /**
+     * List of tasks to execute during each visit.
+     * @private @const {!Array<!Visitor>}
+     */
+    this.visitors_ = [];
+
+    /** @private @const {!Window} */
+    this.win_ = win;
+  }
+
+  /**
+   * Returns a list of visitors that have not yet been marked completed.
+   * @return {!Array<!Visitor>}
+   * @private
+   */
+  getActiveVisitors_() {
+    return this.visitors_.filter(visitor => !visitor.isComplete());
+  }
+
+  /**
+   * @param {!VisitorCallbackTypeDef} callback
+   * @param {number=} maxAncestorsToVisit The limit of how many ancestors this
+   *   task should be executed on. Must be positive.
+   * @return {!DomAncestorVisitor}
+   */
+  addVisitor(callback, maxAncestorsToVisit = 100) {
+    this.visitors_.push(new Visitor(callback, maxAncestorsToVisit));
+    return this;
+  }
+
+  /**
+   * @param {?Element} element
+   */
+  visitAncestorsStartingFrom(element) {
+    let el = element;
+    let visitors = [];
+    while (el && (visitors = this.getActiveVisitors_()).length) {
+      const style = computedStyle(this.win_, el);
+      visitors.forEach(visitor =>
+        visitor.callback(dev().assertElement(el), style)
+      );
+      el = el.parentElement;
+    }
+    this.visitors_.forEach(visitor => visitor.setComplete());
+  }
+}
+
+class Visitor {
+  /**
+   * @param {!VisitorCallbackTypeDef} callback
+   * @param {number} maxAncestorsToVisit
+   */
+  constructor(callback, maxAncestorsToVisit) {
+    devAssert(
+      maxAncestorsToVisit > 0,
+      'maxAncestorsToVisit must be a positive value.'
+    );
+
+    /** @private @const {!VisitorCallbackTypeDef} */
+    this.callback_ = callback;
+
+    /** @private {number} */
+    this.maxAncestorsToVisit_ = maxAncestorsToVisit;
+
+    /** @private {boolean} */
+    this.complete_ = false;
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {!Object<string, string>} style
+   */
+  callback(element, style) {
+    devAssert(
+      !this.complete_,
+      'Attempted to execute callback on completed visitor.'
+    );
+    let result;
+    try {
+      result = this.callback_(element, style);
+    } catch (e) {
+      dev().warn(
+        'DOM-ANCESTOR-VISITOR',
+        `Visitor encountered error during callback execution: "${e}".`
+      );
+    }
+    if (!--this.maxAncestorsToVisit_ || result != undefined) {
+      this.complete_ = true;
+    }
+  }
+
+  /** @return {boolean} */
+  isComplete() {
+    return this.complete_;
+  }
+
+  /** Marks the visitor as complete. */
+  setComplete() {
+    this.complete_ = true;
+  }
```

Comment:
do get/set complete need to be public, or can DomAncestorVisitor just use complete_ directly?

Reply:
Done.